### PR TITLE
deep book order query

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

Added new order_query module to deepbook for querying paginated orders.

2 new APIs have been added:

```
public fun iter_bids<T1, T2>(
         pool: &Pool<T1, T2>,
         // tick level to start from
         start_tick_level: Option<u64>,
         // order id within that tick level to start from
         start_order_id: Option<u64>,
         // if provided, do not include orders with an expire timestamp less than the provided value (expired order),
         // value is in microseconds
         min_expire_timestamp: Option<u64>,
         // do not show orders with an ID larger than max_id--
         // i.e., orders added later than this one
         max_id: Option<u64>,
): OrderPage
```

```
public fun iter_asks<T1, T2>(
         pool: &Pool<T1, T2>,
         // tick level to start from
         start_tick_level: Option<u64>,
         // order id within that tick level to start from
         start_order_id: Option<u64>,
         // if provided, do not include orders with an expire timestamp less than the provided value (expired order),
         // value is in microseconds
         min_expire_timestamp: Option<u64>,
         // do not show orders with an ID larger than max_id--
         // i.e., orders added later than this one
         max_id: Option<u64>,
): OrderPage
```
## Test Plan 

Unit tests and manually local tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/deepbook/tests/order_query_tests.move [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    const BOB: address = @0x1337;
+
+    #[test]
+    fun test_order_query_pagination() {
+        let scenario = prepare_scenario();
+        add_orders(200, TIMESTAMP_INF, none(), &mut scenario);
+        let pool = test_scenario::take_shared<Pool<SUI, USD>>(&mut scenario);
+        let page1 = iter_bids(&pool, none(), none(), none(), none());
+        assert!(vector::length(order_query::orders(&page1)) == 100, 0);
+        assert!(order_query::has_next_page(&page1), 0);
```
Comment (healthydeve) [inline crates/sui-framework/packages/deepbook/tests/order_query_tests.move:38]: does this pagination change if we insert orders in the middle of paginating?

File: crates/sui-framework/packages/deepbook/sources/order_query.move [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        next_order_id: Option<u64>
+    }
+
+    // return an OrderPage starting from start_tick_level + start_order_id
+    // containing as many orders as possible without going over the
+    // dynamic fields accessed/tx limit
+    public fun iter_bids<T1, T2>(
+        pool: &Pool<T1, T2>,
+        // tick level to start from
+        start_tick_level: Option<u64>,
```
Comment (healthydeve) [inline crates/sui-framework/packages/deepbook/sources/order_query.move:32]: Can we define this a little more clearly in comments, i.e. whether or not the start ticks start from the bottom of bid list (lowest priced bids, or from the top of the book highest priced bids).

In the codes case it looks like its going from bottom to top of bids.

File: crates/sui-framework/packages/deepbook/sources/order_query.move [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            has_next_page,
+            next_tick_level,
+            next_order_id
+        }
+    }
+
+    public fun iter_asks<T1, T2>(
+        pool: &Pool<T1, T2>,
+        // tick level to start from
+        start_tick_level: Option<u64>,
```
Comment (healthydeve) [inline crates/sui-framework/packages/deepbook/sources/order_query.move:62]: same comment here for asks

File: crates/sui-framework/packages/deepbook/tests/order_query_tests.move [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    const CLIENT_ID_ALICE: u64 = 0;
+    const FLOAT_SCALING: u64 = 1000000000;
+    const CANCEL_OLDEST: u8 = 0;
+    const TIMESTAMP_INF: u64 = ((1u128 << 64 - 1) as u64);
+
+    const OWNER: address = @0xf;
+    const ALICE: address = @0xBEEF;
+    const BOB: address = @0x1337;
+
+    #[test]
```
Comment (healthydeve) [inline crates/sui-framework/packages/deepbook/tests/order_query_tests.move:31]: For sanity test case, can we have an explicit test case where we insert orders below a certain tick size, and then try to query for a range that does not include that tick size? This is a current bug we have in deepbook and I'd like a test for this.

File: crates/sui-framework/packages/deepbook/sources/order_query.move (hunk 2 above)
Comment (patrickkuo) [reply crates/sui-framework/packages/deepbook/sources/order_query.move:32]: I am going to add a ascending bool arg to allow user to choose the order of the orders (lol)

File: crates/sui-framework/packages/deepbook/tests/order_query_tests.move (hunk 1 above)
Comment (patrickkuo) [reply crates/sui-framework/packages/deepbook/tests/order_query_tests.move:38]: the orders are ordered by (tick, order id), so adding a order will change the page, however user can use the max_id filter to ignore any new additions.

**Review by healthydeve — APPROVED** [review]

**Pull request merged**